bug: handle subpixel values in viewport dims

les2 opened the pull request:
Rounds the dimensions on the DOMRect of the in-viewport element since
`jQuery.innerWidth()` (or `height`) return integers.
`Element.getBoundingClientRect()` can return subpixels (as floating
point numbers) that if not handled will result in elements not being
considered in the viewport by the `is-in-viewport` helper (even though
they are).

This is particularly important when the user has zoomed the window,
in which case the browser may scale the size of the viewport, resulting
in the dimensions having fractional parts, e.g., the width might be
`1451.1112060546875`.

Closes #96

# Details

Rounds the `right` and `bottom` properties on the `DOMRect` returned from `Element.getBoundingClientRect()` on the element being checked to integers.


# Functional Testing

I have tested by `npm link`ing this change against [this repo](https://les2.github.io/ember-light-table-zoom-bug/#/elt-demo) and verifying that the viewport detection works correctly.

I have personally tested in recent versions the following browsers on OS X El Capitan using the above method:
- Firefox
- Safari
- Chrome

les2 commented:
cc @offirgolan

offirgolan commented:
@les2 thanks again for seeing this issue through. The change makes sense and looks good to me but it seems as though tests are failing.

offirgolan commented:
> Will not support IE<=8 because the width/height properties are not available. Do we care about that?
There may be other problems with IE<=8 --- I don't even have an IE testing VM around anymore!

I think we should fallback to the default implementation using jQuery innerWidth and innerHeight if getBoundingClientRect isnt there. That way this fix is backwards compatible for people still wanting to support IE<=8

les2 commented:
will update shortly.

i think i forgot to rebase away a failing test i added which does not apply
to my implementation.

On Sun, Dec 4, 2016, 6:09 PM Offir Golan <notifications@github.com> wrote:

> Will not support IE<=8 because the width/height properties are not
> available. Do we care about that?
> There may be other problems with IE<=8 --- I don't even have an IE testing
> VM around anymore!
>
> I think we should fallback to the default implementation using jQuery
> innerWidth and innerHeight if getBoundingClientRect isnt there. That way
> this fix is backwards compatible for people still wanting to support IE<=8
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DockYard/ember-in-viewport/pull/97#issuecomment-264739347>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBEYBbUSniTUQfc8NwCf9-z2-8tveUKks5rE0gUgaJpZM4LDpi6>
> .
>

les2 commented:
Not sure what's going on with tests.

- on master branch, acceptance test hangs for me
- on this PR branch, PhantomJS hangs but Chrome passes (only Chrome is run in CI)

How do you ask travis to run the tests again?

les2 commented on addon/mixins/in-viewport.js:
```diff
@@ -91,14 +91,23 @@ export default Mixin.create({
+    if (!('width' in contextBoundingClientRect)) {
+      // IE8 and lower:
+      width = contextBoundingClientRect.right - contextBoundingClientRect.left;
+      height = contextBoundingClientRect.bottom - contextBoundingClientRect.top;
```
@offirgolan I think this will work to compute the `width` and `height` in IE8. `width` and `height` are in a sense redundant properties on the DOMRect as they can be derived from `right - left` and `bottom - top`, respectively.

Apparently this is quite messy. This article from 2008 by John Resig makes me think its not this simple: http://ejohn.org/blog/getboundingclientrect-is-awesome/

Modern jQuery deals with some of this mess here: https://github.com/jquery/jquery/blob/master/src/offset.js#L78-L118

offirgolan replied:
@les2 I'm not too certain as I have no way to test IE on my machine. I feel like the final decision here should be left up to @poteto. 

I think that if this isnt a for sure solution, then the best thing might be to use `contextBoundingClientRect.{width,height}` and default back to the previous jQuery solution for older browsers.

les2 replied:
@offirgolan implemented your suggestion.

note: this does mean that IE<=8 users will get the rounding bug.

there is an alternative implementation that I probably should have pushed first: just round everything to whole numbers with `Math.round` in the `isInViewport` function. the downside is that you lose subpixel accuracy with the in-viewport detection. but 99.999% of users probably don't care about it.

To make it clearer, I could remove these commits and change `isInViewport` as follows (just adding the `Math.round`):
```
export default function isInViewport(boundingClientRect = {}, height = 0, width = 0, tolerance = defaultTolerance) {
  const { top, left, bottom, right } = boundingClientRect;
  const tolerances = assign(defaultTolerance, tolerance);
  const {
    top: topTolerance,
    left: leftTolerance,
    bottom: bottomTolerance,
    right: rightTolerance
  } = tolerances;

  return (
    (Math.round(top) + topTolerance)       >= 0 &&
    (Math.round(left) + leftTolerance)     >= 0 &&
    (Math.round(bottom) - bottomTolerance) <= height &&
    (Math.round(right) - rightTolerance)   <= width
  );
}
``` 

Thoughts on this? One advantage is that it is really easy to test as this is pure function and there's an existing unit test for other cases. (This was actually my first thought).

@poteto thoughts?

les2 commented:
@offirgolan @poteto All checks passing / all feedback received so far is addressed. Please let me know if there's something further I can clean up.

les2 commented:
@kellyselden thoughts on this one? pinging you since you are the second most active contrib. here next to `@poteto`

les2 commented:
OOOPS! Just found a huge bug! I'll fix shortly. It got added when the `if-else` was added for fallback behavior! cc @offirgolan

les2 commented:
Okay - fixed bug! ready for review again.

poteto commented:
Tests are failing, can you fix please?

les2 commented:
will take a look when i get to a computer.

On Tue, Dec 13, 2016, 2:24 PM Lauren Tan <notifications@github.com> wrote:

> Tests are failing, can you fix please?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/DockYard/ember-in-viewport/pull/97#issuecomment-266835636>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBEYGTHjUr5Pd4ueN4afGNHsUXWJ9qjks5rHvDigaJpZM4LDpi6>
> .
>

kellyselden commented:
It doesn't look like this addon is targeting Ember 1.x, and since IE 8 was dropped in Ember 2.x, should we not include IE8 compat code in here?

les2 commented:
> It doesn't look like this addon is targeting Ember 1.x, and since IE 8 was dropped in Ember 2.x, should we not include IE8 compat code in here?

I don't disagree.

offirgolan commented:
@kellyselden Although this addon doesnt target Ember 1.x it doesnt mean that there aren't apps and addons using Ember 1.x with this addon. 

If this addon doesnt support Ember 1.x, then publishing this change would have to result in a major release and state that this addon no longer supports Ember 1.x. 

The implemented IE8 fallback is literally 2 lines of code so I honestly dont think it should cause much headache. I guess the call here is on the DockYard team to decide what they would perfer.

les2 commented:
@offirgolan @kellyselden @poteto I have changed the implementation (mainly to get the tests to pass) and the following is different:
- it doesn't change the browser compatibility (it doesn't touch that code)
- these new changes are testable and i include some unit tests

I have updated the commit message / PR description to match.

Please review again when you can.

les2 commented on addon/utils/is-in-viewport.js:
```diff
@@ -22,7 +22,7 @@ export default function isInViewport(boundingClientRect = {}, height = 0, width
-    (bottom - bottomTolerance) <= height &&
-    (right - rightTolerance)   <= width
+    (Math.round(bottom) - bottomTolerance) <= Math.round(height) &&
+    (Math.round(right) - rightTolerance)   <= Math.round(width)
```
@offirgolan Just rounding it here. This appears to be the only change needed!

offirgolan replied:
Why is there only a need to round `right` and `bottom`?

les2 replied:
the other properties are only `foo >= 0` .... as long as it's positive `foo`, no need to round. even if it's a smash fraction like 1e-100

offirgolan replied:
Ah right! That one totally went over my head 😝

les2 commented:
ping @poteto.

les2 commented:
Happy New Year, @poteto .

anything you want to see changed? I've deployed a patch for this in production for work and have seen no problems. 

would like for other users of this addon and ELT to enjoy this bug fix, too.

I'm happy to make any updates folks think are necessary.

kellyselden commented:
LGTM Thanks for all the work. @les2 @offirgolan

kellyselden commented:
Released in v2.1.1
